### DISCUSSION TITLE
Multi-Buy promotion not working

### DISCUSSION THREAD

**Discussion by BenderNETs** [openstore-ecommerce/OpenStore#185] [General]
Trying to use the multi buy promotion option. I created a new promotion property "Discountable" and assigned the property to all my products. I added a new promo "Buy 4 get 1 free" and selected the new property "Discountable" and a quantity of 4. In the amount I put in the amount of 5.50 (which is the prize of all my products). I also added a valid date from/until. When I select a product in the shop I can see the message from the promo "Buy 4 get 1 free", but when I select the quantity to 5 or more and click on the "Buy" button I don't see the correct prize in the cart. I tried different options (select more than 4 products of one product, select one product of 5 different products etc.) but none are working. Any ideas?

**Comment by DNNMonster** [comment]
Did you go through the Scheduler Setup? 

https://doc.openstore-ecommerce.com/

 It'll call the DoWork function on the MultiBuyPromoScheduler which should update your prices.  

https://github.com/openstore-ecommerce/OpenStore/blob/ce09324695cc817ceb49919fa0dfc902950595ea/Providers/PromoProvider/PromoProvider.cs#L31

**Comment by DNNMonster** [comment]
Sorry.  I spoke too soon.  I don't use this feature but I tried and ran into some issues.

It looks like we may need to overload GetProductData with the portalid so we can inject this into a new ctor for ProductData.  Otherwise we fail at line 52 in the ProductData class since there won't be any PortalSettings.Current.PortalId when the scheduler executes. 

Dave may have some insight here.  It's possible I'm overlooking something.

**Reply by DNNMonster** [reply]
Double spoke too soon.  Actually the ProductData constructor to support passing a portalid is in place.

**Comment by BenderNETs** [comment]
Not quite sure why would I need a scheduler for this feature? If I setup a promo and say if you buy 5 items you only pay for 4.
So the customer clicks on a product, changes the quantity to to 5 and clicks on the "Buy" button. Right after you should see 5 items in the shopping basket and the total amount of $22.00 (item prize is $5.50). Currently I see $27.50 which is not correct. Even when go through the order process I can't see any discount. I see that the shop knows that the product is a promo product because the message I setup in the promo shows up on the product page saying "Buy 4 get 1 free".
I see the message at the bottom in the promo setup: "Promotions are calculated by the scheduler. Ensure the NBS Scheduler is running."
I added a new scheduler according to the documentation with no luck. The scheduler runs successfully but the problem I mentioned is still there.

**Reply by DNNMonster** [reply]
As far as why the scheduler was involved I'm not exactly sure but it's not a blocking issue.   

I think the issue is that we need to improve our usage of the portalid so that out of http context calls don't fail.

I'm looking at it.

**Comment by DNNMonster** [comment]
I think the fixes for the portal id usage are here.  If you have you're dev environment up you can give it a test drive.  I'll PR it once it's marinated a couple of days.

https://github.com/DNNMonster/Open-Store/tree/bugfix/portal-id-usage

**Comment by leedavi** [comment]
The scheduler is needed because it processors on a bulk update method.

This was a plugin my client wanted and at the time I found a problem with speed.  It calculates the prices and saves them to the product data.  So all data required is already calculated.

It's not the best way, but it works.

So, basically you need the scheduler running.

https://doc.openstore-ecommerce.com/#11498

**Comment by leedavi** [comment]
P.S.  You should always have the scheduler running, it cleans up old carts in the DB.

**Comment by BenderNETs** [comment]
The scheduler is running. Here's the log:

OK Service Start. Failed. providername: NBrightBuy.PromoProvider Nevoweb.DNN.NBrightBuy.Providers.PromoProvider.MultiBuyPromoScheudler Portal:LHF Designs Error:System.NullReferenceException: Object reference not set to an instance of an object. at Nevoweb.DNN.NBrightBuy.Components.ProductData.ResetData(Int32 productId, String lang, Boolean hydrateLists, String typeCode) at Nevoweb.DNN.NBrightBuy.Components.ProductUtils.GetProductData(Int32 productId, String lang, Boolean hydrateLists, String typeCode) at Nevoweb.DNN.NBrightBuy.Providers.PromoProvider.PromoUtils.CalcMultiBuyPromoItem(NBrightInfo p) at Nevoweb.DNN.NBrightBuy.Providers.PromoProvider.PromoUtils.CalcMultiBuyPromo(Int32 portalId) at Nevoweb.DNN.NBrightBuy.Components.NBrightBuyScheduler.DoWork()

**Comment by leedavi** [comment]
I've just checked one of our clients website and cannot see any problem with the promotions.

This does look like data issue.


Make sure you have some promotions setup, I do notice the blank entry at the top of the image... not sure why, but maybe you need a minimum.

![image](https://user-images.githubusercontent.com/5583141/131337268-270bf676-4ef1-4db2-bd65-c897721cd176.png)


What version of OpenStore and DNN are you using?

**Reply by leedavi** [reply]
The version I have checked is v3.6.  We don't have this plugin running on DNN9 and OpenStore v4.  Maybe there is some issue on DNN9.

**Reply by DNNMonster** [reply]
It may be version specific.  I am not testing against v3 so I can't speak to it.  

I also see this issue on DNN 9.10.0 with OS 4.1.5

I think PR #189 should address this.  Let me know if there is an issue with the code.

**Comment by BenderNETs** [comment]
I'm running DNN 9.7.0 with OpenStore 4.1.4
Currently I have only one promotion under Multi-Buy (Vouchers and Group are empty).

Here are some screenshots:

![image](https://user-images.githubusercontent.com/17228217/131353533-75e2181e-4d5f-45b8-9f96-81ffbff37494.png)

![image](https://user-images.githubusercontent.com/17228217/131353589-5b8b8323-a113-4e6d-8076-339de16abc9e.png)

![image](https://user-images.githubusercontent.com/17228217/131353640-486e26de-d66e-4b6b-9549-6c71dabaedee.png)

![image](https://user-images.githubusercontent.com/17228217/131353719-67af1eed-5687-49c9-9388-1cd75c37814a.png)

![image](https://user-images.githubusercontent.com/17228217/131353823-d3da8116-60f0-445c-90bf-8035462d12be.png)

**Comment by leedavi** [comment]
I think @DNNMonster has fixed this for you.  I have merged the change into the next version.

@DNNMonster may have a install package you can use before we release the official version.

**Comment by DNNMonster** [comment]
No unofficial releases are happening. If you want to get ahead of the official release cycle then it's best if you fork the project. 

Then you could compile and also help contribute if something is in your wheelhouse.  All open source projects can always use extra help!

**Comment by BenderNETs** [comment]
As mentioned I'm running DNN 9.7.0 with OpenStore 4.1.4. I did not want to upgrade to the latest version, so I just made the changes with the missing portalid and recompiled 4.1.4. The job runs now without any exceptions. I guess the issue running the scheduled job is resolved, but the calculation is still not working. When I have 5 items of the same product in my cart it still shows the price of 5 and not 4. As mentioned, on the details page I can see the promo message for the product.

I looked at the CalculatePromotion method in the PromoProvider. Inside the loop there is this line:
if (cartItemInfo.GetXmlProperty("genxml/productxml/genxml/hidden/promotype") == "PROMOMULTIBUY")

Problem is that the current CART xml does not even have "genxml/productxml/genxml/hidden/promotype" and that's why the calculation doesn't work.

**Comment by DNNMonster** [comment]
Can you verify that your dates are valid (your screen shot looks good but it may not match your current state).   You'll see a date check that wraps a call to the function that should be setting the promotype value.  It's possible this is affecting you.

If the dates look good then you could put a breakpoint in the FlagProductMultiBuy function and help us get to the root problem.

My install I tested against worked as expected for the percentage based discounts.   I don't think I tested the amount based so it's possible I missed something.  I will re-visit it if you still see issues after a date check. 


https://github.com/openstore-ecommerce/OpenStore/blob/60da642a1be491bfc49d913e5baa6f6d33da30ce/Providers/PromoProvider/PromoUtils.cs#L412

**Comment by BenderNETs** [comment]
I think the dates look good.

`<genxml>
  <files />
  <hidden>
    <lastcalculated>2021-08-31T21:00:30.8628914-06:00</lastcalculated>
  </hidden>
  <textbox>
    <name>Buy 4 Get 1 free</name>
    <buyqty>4</buyqty>
    <amount datatype="double">5.5</amount>
    <validfrom datatype="date">2021-08-20T00:00:00</validfrom>
    <validuntil datatype="date">2021-09-30T00:00:00</validuntil>
  </textbox>
  <checkbox>
    <disabled>False</disabled>
  </checkbox>
  <dropdownlist>
    <propbuy selectedtext="">1351</propbuy>
  </dropdownlist>
  <checkboxlist />
  <radiobuttonlist>
    <amounttype>1</amounttype>
  </radiobuttonlist>
</genxml>`

**Comment by BenderNETs** [comment]
I don't see any promo node in the product xml
"genxml/hidden/promoid"  (missing)
"genxml/hidden/promoname"  (missing)
"genxml/hidden/promotype"  (missing)
"genxml/hidden/promocalcdate"  (missing)

**Comment by DNNMonster** [comment]
I believe this may be a math confusion issue related to our help text which we can improve upon.  For amount based group discounts the help text should read "amount discounted from the unit cost"

If you view the code for amount based discounts and you'll see that the discount amt that get's applied is using a formula which takes the unit cost of an item and then subtracts the input value which you enter into the promo configuration.

https://github.com/openstore-ecommerce/OpenStore/blob/60da642a1be491bfc49d913e5baa6f6d33da30ce/Providers/PromoProvider/PromoProvider.cs#L78

If I understand things correctly, in your situation you're product has a unit price of 5.50 and you're amount entered into your input for the promo configuration discount amount is also 5.50.

I believe this results in an actual applied discount amount of 0.

I wanted to suggest trying a 100% off discount for a quantity above 4.  Then adding a max qty for the product of 5.  Unfortunately I am seeing an issue with this max qty not being enforced so it won't work just yet.  Someone will inevitably come and try to buy your entire stock for the price of 4 units.

For now you may have to consider adjusting you're offering to something that works with the supported case scenarios,  Buy 4 items and get $2 off each additional item.  Or Buy 4 and get 10% off each additional item.

You should be able to verify this by altering your offering to test a supported multi-buy discount scenario, run the OS scheduler and making sure once again that the math matches expectations.

If the numbers are still wrong we'll keep digging.

**Comment by BenderNETs** [comment]
I changed the discount from amount to percentage, but it is still not working. When I delete the whole promotion, I can still see the message  "If you buy 4 items you will get 1 item free" for each product on the details page. I already ran the NBS Scheduler multiple times and also reset the whole site. Is this a known issue or did I do anything wrong?

![image](https://user-images.githubusercontent.com/17228217/131902651-15894cf8-e31f-4237-a449-82e319b5f9de.png)

![image](https://user-images.githubusercontent.com/17228217/131902715-1bfb2b19-5a84-41c9-aa6d-1442d22e8d9c.png)

![image](https://user-images.githubusercontent.com/17228217/131902935-6e46412d-f6de-4e30-96a9-e537341beadf.png)

**Comment by DNNMonster** [comment]
I am not able to reproduce the error so it's hard to say.  

Both amount based and percentage based multi-buy promotions work as expected on the installation I'm testing against.

**Comment by leedavi** [comment]
The calculated prices are saved on the "PRD" type record.  Can you soo them and are they OK.  it could be a template related issue.

**Reply by leedavi** [reply]
This should look something like this.....

![image](https://user-images.githubusercontent.com/5583141/132147911-90dd53d2-54fb-41a2-816d-028a4a56d5c0.png)

**Comment by BenderNETs** [comment]
The only item missing is "defaultcadid"

![image](https://user-images.githubusercontent.com/17228217/132241403-0a3d9a2f-bfff-4d08-b762-95df15ff109b.png)

**Comment by leedavi** [comment]
Do you have any scheduler errors in the DNN logs?

**Comment by BenderNETs** [comment]
No errors in the logs
![image](https://user-images.githubusercontent.com/17228217/132344800-cccc9376-9550-43c2-9591-12110eb6e801.png)

**Comment by leedavi** [comment]
Can you check the event log in DNN.  If the scheduler finished, it will say OK even if a plugin failed.  The errors are recorded in the event log.

**Comment by BenderNETs** [comment]
I checked the DNN event logs too. There is nothing. The last error I can see was about a week ago (which I posted earlier) before I included the fix. I removed the promotion "MultiBuy" a few days ago as mentioned before and ran into the issue that the promotion description was not removed. I ended up to manually remove the xml item through SQL for all products.
